hs_labels/models.py

```diff
@@ -0,0 +1,876 @@
+
+        * 1 or LabelCodes.LABEL:
+            a regular label for the object.
+
+        * 2 or LabelCodes.FOLDER:
+            a folder for the object.
+
+        * 3 or LabelCodes.FAVORITE:
+            marks the object as a favorite.
+
+        * 4 or LabelCodes.MINE:
+            marks the object to be included in my resources
+
+        * 5 or LabelCodes.SAVEDLABEL:
+            creates a label that can be used later for labeling a resource
+
+    """
+    LABEL = 1
+    FOLDER = 2
+    FAVORITE = 3
+    MINE = 4
+    SAVEDLABEL = 5
+    LABEL_CHOICES = (
+        (LABEL, 'Label'),
+        (FOLDER, 'Folder'),
+        (FAVORITE, 'Favorite'),
+        (MINE, 'Mine'),
+        (SAVEDLABEL, 'Saved Label')
+    )
+
+####################################
+# user access to resources
+####################################
+
+
+class UserResourceLabels(models.Model):
+    """ Labels of a user for a resource
+
+    This model stores labels of an individual user, like an access
+    control list; There are several kinds of labels documented in LabelCodes.
+    These are similar in implementation but differ in semantics.
+    """
+
+    kind = models.IntegerField(choices=LabelCodes.LABEL_CHOICES,
+                               editable=False,
+                               default=LabelCodes.LABEL)
+    # provenance of label start
+    start = models.DateTimeField(editable=False, auto_now=True)
+
+    # if end is non-null, then the label is no longer valid. Not absolutely sure we need this.
+    # end  = models.DateTimeField(editable=False, null=True, auto_now=False)
+
+    # Utilize an underlying class of User to hold labels for resources.
+    # This creates an implicit query User.ulabels that represents all labels
+    # for a user.
+    ulabels = models.ForeignKey('UserLabels', null=False, editable=False,
+                                related_name='ul2url',         # unused but must be defined and unique
+                                help_text='user assigning a label')
+    # There is a basic design question as to whether the target of a label should be a
+    # Resource proper, or an annotation class that contains mainly methods for the resource.
+    # Should this be 'BaseResource' or 'ResourceLabels'?
+    rlabels = models.ForeignKey('ResourceLabels', null=True, editable=False,
+                                related_name="rl2url",     # unused but must be defined and unique
+                                help_text='resource to which a label applies')
+    label = models.TextField()
+
+
+class UserLabels(models.Model):
+    """
+    Projection class puts methods and content inside basic User object
+    so that one can access things easily from that context.
+    This model is injected into the BaseResource as the related name "ulabels".
+    Thus for an User u, u.ulabels is this model.
+    """
+
+    user = models.OneToOneField(User,
+                                editable=False,
+                                null=True,
```

Comment:
Actually, this still doesn't follow. If we need users to save labels without resources, doesn't that mean the key to the **resource** may be null but we'd still always need a key to the **user**?